core-rust-bridge/src/main/java/com/radixdlt/statecomputer/ProtocolState.java

```diff
@@ -0,0 +1,121 @@
+ * caused by bugs or programming or logic errors or otherwise);
+ *
+ * A. any offer, purchase, holding, use, sale, exchange or transmission of any
+ * cryptographic keys, tokens or assets created, exchanged, stored or arising from any
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+package com.radixdlt.statecomputer;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.radixdlt.lang.Option;
+import com.radixdlt.lang.Tuple;
+import com.radixdlt.protocol.ProtocolUpdate;
+import com.radixdlt.protocol.ProtocolUpdateEnactmentCondition;
+import com.radixdlt.sbor.codec.CodecMap;
+import com.radixdlt.sbor.codec.EnumCodec;
+import com.radixdlt.sbor.codec.StructCodec;
+import com.radixdlt.utils.UInt64;
+
+public record ProtocolState(
+    Option<UInt64> currentEpoch,
+    String currentProtocolVersion,
+    ImmutableMap<UInt64, String> enactedProtocolUpdates,
+    ImmutableList<UnenactedProtocolUpdate> unenactedProtocolUpdates) {
+
+  public static void registerCodec(CodecMap codecMap) {
+    codecMap.register(
+        ProtocolState.class,
+        codecs -> StructCodec.fromRecordComponents(ProtocolState.class, codecs));
+
+    codecMap.register(
+        UnenactedProtocolUpdate.class,
+        codecs -> StructCodec.fromRecordComponents(UnenactedProtocolUpdate.class, codecs));
+
+    codecMap.register(
+        UnenactedProtocolUpdateState.class,
+        codecs ->
+            EnumCodec.fromPermittedRecordSubclasses(UnenactedProtocolUpdateState.class, codecs));
+
+    codecMap.register(
+        UnenactedProtocolUpdateState.SignalledReadinessThresholdState.class,
+        codecs ->
+            StructCodec.fromRecordComponents(
+                UnenactedProtocolUpdateState.SignalledReadinessThresholdState.class, codecs));
+  }
+
+  public record UnenactedProtocolUpdate(
```

Comment:
I'm not quite following this structure. Why the need for `UnenactedProtocolUpdateState` if `ProtocolUpdate` already has a condition?

Reply:
(as discussed separately) this also keeps the state of a condition